### DISCUSSION TITLE
Is it possible to change product folder name in pomless build of the Mac artifact?

### DISCUSSION THREAD

**Discussion by guw** [eclipse-tycho/tycho#4424] [Q&A] [answered]
I have a Tycho pomless product build. The products are generated without problems. The product definition declares a custom launcher name. This works, i.e. Tycho creates the product for all platforms using the custom launcher name.

The MacOS artifacts are packaged as `Eclipse.app` app bundles. Is it possible to disable this in a pomless Tycho build or modify the app bundle name to something else?

Also it looks like all other platforms don't use a root folder whereas MacOS artifacts are the only ones using `Eclipse.app`.

**Comment by laeubi** [comment]
> The MacOS artifacts are packaged as `Eclipse.app` app bundles. Is it possible to disable this in a pomless Tycho build or modify the app bundle name to something else?

You can use all configuration options in your configurator-pom to customize this e.g. 
https://tycho.eclipseprojects.io/doc/latest/tycho-p2-director-plugin/materialize-products-mojo.html

but I think the name of the folders (e.g. `Eclipse.app`) is required by mac itself (@sratz ?) so it seems not configurable yet. Can you explain what exactly is the desired outcome here, so what specifically should be disabled?

**Reply by guw** [reply]
I found the explanation for what I am observing:
https://github.com/eclipse-tycho/tycho/blob/85e1103adf8d79ef69d6eee8deb1be8ed840a28f/tycho-p2-director-plugin/src/main/java/org/eclipse/tycho/plugins/p2/director/Product.java#L105-L112

AFAIK the logic there is correct, i.e. appending `.app` on MacOS. 

A property `pom.model.property.rootFolder = foo` would be nice but that's a feature request. How do I set the configurator pom? Would that be just the "parent" pom?

**Reply by laeubi** [reply, marked as answer]
Yes you can configure it in the `pluginManagement` section of the parent. There you can of course also use a property (that then can be overwritten in `build.properties` with `pom.model.property.myRootFolder = foo` this is quite common to use and make pomless more configurable.

**Reply by guw** [reply]
Thanks Christoph! I'll give that a try.

**Reply by guw** [reply]
@HannesWell Did you implement the properties injection mentioned in https://github.com/eclipse-tycho/tycho/issues/913#issuecomment-1126680550?

**Reply by HannesWell** [reply]
Not yet, unfortunately.
But if you have only one *.product file in a folder you can probably inject the properties 'manually' by creating a build.properties file next to the product file and adding corresponding entries like:
`pom.model.property.product-id  = my-cool-product`

For more information see https://github.com/eclipse-tycho/tycho/wiki/Tycho-Pomless#define-properties

**Reply by merks** [reply]
This is what Oomph does:

https://github.com/eclipse-oomph/oomph/blob/e8303cc893b0f41aa09c0856da89ab03ff4458ee/products/org.eclipse.oomph.setup.installer.product/pom.xml#L367-L373

**Reply by sratz** [reply]
macOS app bundles need to follow a certain layout: https://developer.apple.com/library/archive/documentation/CoreFoundation/Conceptual/CFBundles/BundleTypes/BundleTypes.html
One requirement is that the bundle directory ends in `.app`.

The configuration via 
```
      <plugin>
        <groupId>org.eclipse.tycho</groupId>
        <artifactId>tycho-p2-director-plugin</artifactId>
        <version>${tycho-version}</version>
        <configuration>
          <products>
            <product>
              <id>id</id>
              <rootFoldes>
                ..
              </rootFoldes>
            </product>
          </products>
        </configuration>
      </plugin>
```
Is the correct way to define the directory name.

An improvement in Tycho could be to not only derive the launcher / binary name from the launcher name, but also the root folder name.

**Comment by jcompagner** [comment]
we just use this:

buchen/fix-info-plist-maven-plugin: Small Maven plugin to "fix" the
Info.plist file generated by p2 during a Tycho product build.
<https://github.com/buchen/fix-info-plist-maven-plugin>

servoy-eclipse/com.servoy.eclipse.product/pom.xml at master ·
Servoy/servoy-eclipse
<https://github.com/Servoy/servoy-eclipse/blob/master/com.servoy.eclipse.product/pom.xml#L175>

that works fine our app is really Servoy.app not Eclipse.app
